### DISCUSSION TITLE
关于OS的问题

### DISCUSSION THREAD

**Discussion by fanfuna** [peng-zhihui/HelloWord-Keyboard#139] [Q&A]
dynamic工程中创建了几个任务，好像只有ThreadGpioUpdate任务有osDelay(20);。ThreadCtrlLoop任务我没看到延迟的操作，本身优先级高，这样为什么不导致一直占用CPU。所以我有几个疑问，1，OS的操作周期是多少1MS还是更短，在哪里查看呢？2，有些任务没有延迟，这样是否符合规范，还是有别的套路，我理解的有误？

**Comment by xingrz** [comment]
https://github.com/peng-zhihui/HelloWord-Keyboard/blob/d612c6f71d2ee9d0b5a7be1854ea251d3e3903ef/2.Firmware/HelloWord-Dynamic-fw/UserApp/main.cpp#L36

https://github.com/peng-zhihui/HelloWord-Keyboard/blob/d612c6f71d2ee9d0b5a7be1854ea251d3e3903ef/2.Firmware/HelloWord-Dynamic-fw/UserApp/main.cpp#L99

https://github.com/peng-zhihui/HelloWord-Keyboard/blob/d612c6f71d2ee9d0b5a7be1854ea251d3e3903ef/2.Firmware/HelloWord-Dynamic-fw/UserApp/main.cpp#L134

注意看这几个地方。实际上，`ThreadCtrlLoop` 的循环里 take 了一个信号量，每次循环到这里会先锁住。在另一个地方，有一个 5000Hz 的定时器 `timerCtrlLoop` 会以固定间隔 give 这个信号量，让 `ThreadCtrlLoop` 循环继续跑。
